### PR TITLE
Refactor Multipart request body parsing

### PR DESCRIPTION
**Pull request by tonyjhuang**
### Description

Pulls out the multipart parsing logic into a single shared file for gcloud.ts and firebase.ts, makes it more flexible, adds validation and testing

### CODE REVIEW

File: src/emulator/storage/multipart.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  metadataRaw: string;
+  dataRaw: Buffer;
+};
+
+/**
+ * Represents a parsed multipart request body. Request bodies can have an
+ * arbitrary number of parts.
+ */
+type MultipartRequestBody = {
+  dataParts: MultipartRequestBodyPart[];
```
[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/emulator/storage/multipart.ts:18]: Good suggestion on the type, changed.

Re: your second suggestion, that's what `ObjectUploadMultipartData` is. If we ever need to parse multipart bodies for non-upload request contexts, it's easy to reuse the internal types (`MultipartRequestBody`, etc) and expose a new type for other use cases

File: src/emulator/storage/apis/gcloud.ts [hunk 2]
```diff
@@ -199,55 +209,39 @@ export function createCloudEndpoints(emulator: StorageEmulator): Router {
...
+      }
+      throw err;
     }
-
-    const bufferOffset = metadataSegment.length + dataSegmentHeader.length;
-
-    const blobBytes = Buffer.from(bodyBuffer.slice(bufferOffset, -`\r\n${boundary}--`.length));
-
-    const metadata = storageLayer.oneShotUpload(
+    const metadata = JSON.parse(metadataRaw)!;
```
[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/emulator/storage/apis/gcloud.ts:231]: done

File: src/test/emulators/storage/multipart.spec.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+{"contentType":"text/plain"}\r
+--b1d5b2e3-1845-4338-9400-6ac07ce53c1e\r
+Content-Type: text/plain\r
+\r
+hello there!
+\r
+--b1d5b2e3-1845-4338-9400-6ac07ce53c1e--\r
+`);
+
+  describe("#parseObjectUploadMultipartRequest", () => {
```
[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/test/emulators/storage/multipart.spec.ts:19]: This actually seems like the canonical way to declare test case names, see https://mochajs.org/#getting-started, although I see now that I forgot the final `()`

[the remark this answers is not in the document]
Comment (tonyjhuang) [reply src/test/emulators/storage/multipart.spec.ts:19]: Would pref to keep this aligned with the mocha user docs

[the remark this answers is not in the document]
Comment (tohhsinpei) [reply src/test/emulators/storage/multipart.spec.ts]: I see, that's fine then, thanks

File: src/emulator/storage/apis/firebase.ts [hunk 4]
```diff
@@ -211,6 +212,22 @@ export function createFirebaseEndpoints(emulator: StorageEmulator): Router {
     );
   });
 
+  const reqBodyToBuffer = async (req: Request) => {
```
Comment (tohhsinpei) [inline src/emulator/storage/apis/firebase.ts:215]: Address linter warning

**Review by tohhsinpei — APPROVED** [review]

File: src/emulator/storage/apis/firebase.ts (hunk 4 above)
Comment (tonyjhuang) [reply src/emulator/storage/apis/firebase.ts]: done

**Pull request merged**